Mathlib/Topology/Algebra/Category/ProfiniteGrp.lean

```diff
@@ -161,4 +165,169 @@ instance : HasForget₂ ProfiniteGrp Grp where
+  letI : TotallyDisconnectedSpace H := Homeomorph.totallyDisconnectedSpace e.toHomeomorph
+  .of H
+
+/-- A closed subgroup of a profinite group is profinite. -/
+def ofClosedSubgroup {G : ProfiniteGrp}
+    (H : Subgroup G) (hH : IsClosed (H : Set G)) : ProfiniteGrp :=
+  letI : CompactSpace H := isCompact_iff_compactSpace.mp (IsClosed.isCompact hH)
+  of H
+
+end ProfiniteGrp
+
+/-!
+# The projective limit of finite groups is profinite
+
+* `FiniteGrp.limit` : the concretely constructed limit of finite group as a subgroup of the pi-type
+
+* `ofFiniteGrpLimit`: direct limit of finite groups is a profinite group
+
+* Verify that the constructed limit satisfies the universal property.
+-/
+
+section Profiniteoflimit
+
+open TopologicalGroup
+
+universe w w'
+
+variable {J : Type v} [Category.{w, v} J] (F : J ⥤ FiniteGrp.{max v w'})
+
+attribute [local instance] ConcreteCategory.instFunLike ConcreteCategory.hasCoeToSort
+
+instance (j : J) : TopologicalSpace (F.obj j) := ⊥
+
+instance (j : J) : DiscreteTopology (F.obj j) := ⟨rfl⟩
+
+instance (j : J) : TopologicalGroup (F.obj j) := {}
+
+namespace FiniteGrp
+
+/-- Concretely constructing the limit of topological group as a subgroup of the  pi-type. -/
+def limit : Subgroup (Π j : J, F.obj j) where
+  carrier := {x | ∀ ⦃i j : J⦄ (π : i ⟶ j), F.map π (x i) = x j}
+  mul_mem' hx hy _ _ π := by simp only [Pi.mul_apply, map_mul, hx π, hy π]
+  one_mem' := by simp only [Set.mem_setOf_eq, Pi.one_apply, map_one, implies_true]
+  inv_mem' h _ _ π := by simp only [Pi.inv_apply, map_inv, h π]
+
+@[simp]
+lemma mem_limit (x : Π j : J, F.obj j) : x ∈ limit F ↔
+    ∀ ⦃i j : J⦄ (π : i ⟶ j), F.map π (x i) = x j := Iff.rfl
+
+lemma limit_closed_in_pi_type : IsClosed ((limit F) : Set (Π j : J, F.obj j)) := by
+  classical
+  let S ⦃i j : J⦄ (π : i ⟶ j) : Set (Π j : J, F.obj j) := {x | F.map π (x i) = x j}
+  have hS ⦃i j : J⦄ (π : i ⟶ j) : IsClosed (S π) := by
+    simp only [S, ← isOpen_compl_iff, isOpen_pi_iff]
+    rintro x (hx : ¬ _)
+    refine ⟨{i, j}, fun i => {x i}, ?_⟩
+    simp only [Finset.mem_singleton, isOpen_discrete, Set.mem_singleton_iff, and_self,
+      implies_true, Finset.coe_singleton, Set.singleton_pi, true_and]
+    intro y hy
+    simp only [Finset.coe_insert, Finset.coe_singleton, Set.insert_pi, Set.singleton_pi,
+      Set.mem_inter_iff, Set.mem_preimage, Function.eval, Set.mem_singleton_iff,
+      Set.mem_compl_iff, Set.mem_setOf_eq] at hy ⊢
+    rwa [hy.1, hy.2]
+  have eq : limit F = ⋂ (i : J) (j : J) (π : i ⟶ j), S π := by
+    ext x
+    simp only [Subgroup.coeSubtype, Subtype.range_coe_subtype, SetLike.mem_coe, mem_limit,
+      Set.mem_setOf_eq, Set.mem_iInter]
+    tauto
+  exact eq ▸ isClosed_iInter fun i => isClosed_iInter fun j => isClosed_iInter fun π => hS π
+
+instance : CompactSpace (limit F) := isCompact_iff_compactSpace.mp <|
+  IsClosed.isCompact <| limit_closed_in_pi_type F
```

Comment:
```suggestion

variable {J : Type v} [SmallCategory J] (F : J ⥤ ProfiniteGrp.{max v w'})

namespace FiniteGrp

/-- Concretely constructing the limit of topological group as a subgroup of the  pi-type. -/
def limit : Subgroup (Π j : J, F.obj j) where
  carrier := {x | ∀ ⦃i j : J⦄ (π : i ⟶ j), F.map π (x i) = x j}
  mul_mem' hx hy _ _ π := by simp only [Pi.mul_apply, map_mul, hx π, hy π]
  one_mem' := by simp only [Set.mem_setOf_eq, Pi.one_apply, map_one, implies_true]
  inv_mem' h _ _ π := by simp only [Pi.inv_apply, map_inv, h π]

def profiniteGrpToProfinite : ProfiniteGrp ⥤ Profinite where
  obj G := G.toProfinite
  map f := ⟨f, by continuity⟩

instance : CompactSpace (limit F) := inferInstanceAs
  (CompactSpace (Profinite.limitCone (F ⋙ profiniteGrpToProfinite.{max v w'})).pt)
```
This works, for example, because your `limit` is defeq to the limit in `Profinite` (in line with Filippo's comment about `limit_closed_in_pi_type`)

Reply:
So what would need to change if we need smallcategory?(I haven't use it before)

Reply:
If I write as you suggested, do I need to verify the universal property again? (I guess no?)

Reply:
I think your proof will more or less work

Reply:
reconstructed

Reply:
@dagurtomas But don't we have the same result for topological spaces so that we can deduce this one from that?

Reply:
I guess you mean something like this? It broke the proofs of `map_one'` and  `map_mul'` (sorried below), but that shouldn't be too hard to fix.

```lean
/-- Verify that the limit constructed above satisfies the universal property. -/
def LimitConeIsLimit : Limits.IsLimit (LimitCone F) where
  lift cone := {
    toFun := ((Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite.{max v w'})).lift 
      (profiniteGrpToProfinite.mapCone cone)).toFun
    map_one' := sorry
    map_mul' := sorry
    continuous_toFun := ((Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite.{max v w'})).lift 
      (profiniteGrpToProfinite.mapCone cone)).continuous }
```

Reply:
Yes!

Reply:
The `fac` field is then handled by automation, I think the `uniq` field can be easily solved using `(Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite)).uniq` after proving that `profiniteGrpToProfinite` is faithful.

Reply:
Like this:
```lean
uniq cone m h := by
    have : profiniteGrpToProfinite.Faithful := sorry 
    apply this.map_injective 
    simpa using (Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite.{max v w'})).uniq
      (profiniteGrpToProfinite.mapCone cone) (profiniteGrpToProfinite.map m) 
      (fun j ↦ congrArg profiniteGrpToProfinite.map (h j))
``` 

Reply:
Here is the whole thing:

```lean
/-- Verify that the limit constructed above satisfies the universal property. -/
def LimitConeIsLimit : Limits.IsLimit (LimitCone F) where
  lift cone := {
    toFun := ((Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite)).lift
      (profiniteGrpToProfinite.mapCone cone)).toFun
    map_one' := by
      apply SetCoe.ext
      ext j
      exact map_one (cone.π.app j)
    map_mul' := fun _ _ ↦ by
      apply SetCoe.ext
      ext j
      exact map_mul (cone.π.app j) _ _
    continuous_toFun := ((Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite)).lift
      (profiniteGrpToProfinite.mapCone cone)).continuous }
  uniq cone m h := by
    have : profiniteGrpToProfinite.Faithful := sorry
    apply this.map_injective
    simpa using (Profinite.limitConeIsLimit (F ⋙ profiniteGrpToProfinite)).uniq
      (profiniteGrpToProfinite.mapCone cone) (profiniteGrpToProfinite.map m)
      (fun j ↦ congrArg profiniteGrpToProfinite.map (h j))
```

Unfortunately automation didn't kick in properly in the proofs of `map_one` and `map_mul`. Might be because of some missing simp lemmas in the `Profinite`/`CompHaus`/`TopCat` limit API.

Reply:
I got stuck when proving it, can anyone help? I found nothing to use about the multiplication.

Reply:
Thank you, I add the faithful as an instance. 20 lines less than the unmodified ones.